### ISSUE TITLE
Articles / Papers

### ISSUE THREAD

**Issue by steveoni** [systemEng-Learning/simple-wireguard#1]
Source articles and papers to understand wireguard

Add lists of articles/papers as a comment

**Comment by goodyduru** [comment]
[Wireguard Protocol (Simple)](https://www.wireguard.com/protocol/)
[Whitepaper](https://www.wireguard.com/papers/wireguard.pdf)

**Comment by steveoni** [comment]
https://mikkel.hoegh.org/2019/11/01/home-vpn-server-wireguard/

https://bash-prompt.net/guides/wireguard-setup/

**Comment by goodyduru** [comment]
Entire content of [Linux Journey(Networking Nomad section)](https://linuxjourney.com/) really provides a very good background.

**Comment by steveoni** [comment]
> Entire content of [Linux Journey](https://linuxjourney.com/) Networking Nomad section really provides a very good background.

This looks really good

**Comment by goodyduru** [comment]
Thanks, pay special attention to the [Routing](https://linuxjourney.com/lesson/what-is-a-router) section when you get there.

**Comment by goodyduru** [comment]
It turns out that wireguard is a tun device. I've been reading up on them. They are cool. These are some of them:

https://www.kernel.org/doc/Documentation/networking/tuntap.txt
https://backreference.org/2010/03/26/tuntap-interface-tutorial/
https://jvns.ca/blog/2022/09/06/send-network-packets-python-tun-tap/
https://www.gabriel.urdhr.fr/2021/05/08/tuntap/
https://blog.cloudflare.com/virtual-networking-101-understanding-tap
https://tailscale.com/blog/more-throughput

I'm not done reading them though, will do that. I think they do provide a good foundation to understanding how wireguard works.

**Comment by randomstuff** [comment]
@goodyduru, Note that Kernel-mode implementation of Wireguard is **not** TUN-based.

See the [Quickstart](https://www.wireguard.com/quickstart/):

<img src="https://www.wireguard.com/img/walkthrough.gif" />

Other implementations (for example [BoringTun](https://github.com/cloudflare/boringtun)) [might be](https://github.com/cloudflare/boringtun/blob/master/boringtun/src/device/tun_linux.rs).

**Comment by goodyduru** [comment]
@randomstuff, I agree. User-mode implementations tend to be TUN-based. [Wireguard-go](https://github.com/WireGuard/wireguard-go/blob/master/device/tun.go) and [Wireguard-rs](https://git.zx2c4.com/wireguard-rs/about/#architecture) are TUN-based. Thanks for the clarification.

**Comment by goodyduru** [comment]
This [site](https://wiki.archlinux.org/title/WireGuard) helps with setting up wireguard with tunneling.

**Comment by steveoni** [comment]
[scapy](https://github.com/secdev/scapy) can be use to manipulate and analyse the packet and frame received from tuntap interface

**Comment by goodyduru** [comment]
This [repo](https://github.com/DhavalKapil/icmptunnel) helped me understand tunnelling better from a programming standpoint.

**Comment by goodyduru** [comment]
Saw [this](https://tailscale.com/blog/2021-05-life-of-a-packet) after we had completed the project. Would have really been helpful then. Could still be helpful later.
